proxy: Use TPROXY on shared ports for all proxies (DNS, Kafka, Envoy) v2

jrajahalme opened the pull request:
Use fixed set iptables TPROXY rules for all Cilium L7 redirects. The rules are added after the proxy listeners are started. Bpf programs are changed to return without any packet modifications and without an explicit bpf redirect when packets need to go to the host proxy. All listening sockets are set up to accept foreign IP addressing via IP(V6)_TRANSPARENT option.

DNS uses the new experimental abstracted UDP transport for miekg/dns (via cilium/dns). This is necessary so that a fixed set of sockets can be used to send replies with a foreign IP address/port.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6921)
<!-- Reviewable:end -->

joestringer commented on pkg/envoy/server.go:
```diff
@@ -327,14 +330,14 @@ func (s *XDSServer) AddListener(name string, kind policy.L7ParserType, endpointP
 	if kind == policy.ParserTypeHTTP {
 		listenerConf.FilterChains = append(listenerConf.FilterChains, proto.Clone(s.httpFilterChainProto).(*envoy_api_v2_listener.FilterChain))
-		listenerConf.FilterChains[0].Filters[1].ConfigType.(*envoy_api_v2_listener.Filter_Config).Config.Fields["http_filters"].GetListValue().Values[0].GetStructValue().Fields["config"].GetStructValue().Fields["policy_name"] = &structpb.Value{Kind: &structpb.Value_StringValue{StringValue: endpointPolicyName}}
+		// listenerConf.FilterChains[0].Filters[1].ConfigType.(*envoy_api_v2_listener.Filter_Config).Config.Fields["http_filters"].GetListValue().Values[0].GetStructValue().Fields["config"].GetStructValue().Fields["policy_name"] = &structpb.Value{Kind: &structpb.Value_StringValue{StringValue: endpointPolicyName}}
```
Is there a particular reason to comment these out rather than delete the line? Is this trying to document what the default values are for this value?

joestringer replied:
(Similar comments for other commented-out parts in this patch)

jrajahalme replied:
These are to remind that we set these values dynamically. I haven't really documented this anywhere, but with the commented lines one can see what kind of configuration we send to Envoy (after the dynamically added parts, that is).

joestringer commented on pkg/proxy/envoyproxy.go:
```diff
@@ -46,23 +46,16 @@ func createEnvoyRedirect(r *Redirect, stateDir string, xdsServer *envoy.XDSServe
-			return nil, fmt.Errorf("%s: Cannot create redirect, proxy local endpoint has no IP address", r.id)
-		}
-		xdsServer.AddListener(redir.listenerName, r.parserType, ip, r.ProxyPort, r.ingress, wg)
+		xdsServer.AddListener(redir.listenerName, r.parserType, r.ProxyPort, r.ingress, wg)
```
This commit strips out the IPv4/ipv6 address fetch and policy name setting and replaces with something that's implicit. Is the implicit set aware that IPv4/Ipv6 can be individually disabled? I didn't find the corresponding code with a brief search around this area. (Or is this logic basically shifted to the proxy?)

jrajahalme replied:
`ip` here was used as a the policy name to configure on the new listener. Now we deduce the policy name on the proxy side from the IP addresses of the connection, so there is no need to explicitly configure the policy name on the listener.

We already send two copies of the policies to the proxy (named with the IPv4 and IPv6 addresses of the EP). I haven't tested it, but nothing should break if the copy with e.g., IPv6 address as the name is not sent when IPv6 is not configured (or is disabled) (same for IPv4).

joestringer commented on pkg/envoy/server.go:
```diff
@@ -350,22 +353,31 @@ func (s *XDSServer) AddListener(name string, kind policy.L7ParserType, endpointP
+	if count, ok := s.listeners[name]; ok && count > 0 {
+		if count == 1 {
+			delete(s.listeners, name)
+			listenerRevertFunc = s.listenerMutator.Delete(ListenerTypeURL, name, []string{"127.0.0.1"}, wg.AddCompletion())
```
I'm a bit confused both by both the previous implementation and this new implementation here.

This is `RemoveListener()`. To successfully implement the removal, you must delete the name from `s.listeners` and presumably propagate a deletion through XDS to the XDS server (if the count goes to zero).

However, this function here is the `listenerRevertFunc` which I would expect to be called to _undo_ the listener removal in the case that something else goes wrong during policy computation. Instead it seems to be implementing the actual deletion across XDS. Why does the revert of the removal involve removing the listener?

Put another way, shouldn't revert of the removal re-add the listener?

jrajahalme replied:
`s.listenerMutator.Delete()` does the deletion AND returns the revert function to reverse the deletion it did. So the line here executes the `Delete()` function and stored the returned revert function for later use.

joestringer replied:
Ah, right, returning a function from a function you call which returns a function. Makes sense, for some reason I figured this line was creating a lambda expression and not executing the delete.

joestringer commented on pkg/endpointmanager/manager.go:
```diff
@@ -143,7 +144,7 @@ func Lookup(id string) (*endpoint.Endpoint, error) {
 
 	case endpointid.IPv6Prefix:
-		return lookupIPv4(eid), nil
+		return lookupIPv6(eid), nil
```
Is this a bug? Does it need backporting?

jrajahalme replied:
Looks like a bug, but don't know if anyone has been affected by it, so don't know if backporting is needed.

aanm replied:
Please make this in a new PR so we can backport this.

jrajahalme replied:
Done! (and merged already)

joestringer commented on pkg/proxy/kafka.go:
```diff
@@ -41,74 +43,166 @@ const (
+// mutex protects accesses to the configuration resources.
+var mutex lock.RWMutex
+var kafkaListeners map[uint16]*kafkaListener // key: proxy port
+var kafkaRedirects map[uint64]*kafkaRedirect // key: dst port | ingress << 16 | endpoint ID << 32
```
nit: the implementation uses `dstPort | dir<<16 | eID<<32`, this says "ingress".

jrajahalme replied:
`dir` here is an integer transformation of the boolean `ingress`, which Go does not do implicitly... I'm leaning to leaving it as it is.

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -182,6 +185,103 @@ type GCFilter struct {
+		} else {
+			mapname = MapNameAny6Global
+		}
+	}
```
It'd be nice to refactor finding the map based on CtKey out from this function. It should be able to return `NewMap(...)` based on something similar to the `maps()` function in this file.

jrajahalme replied:
Will need to revisit the ctmap lookup functionality to support local conntrack maps as well.

jrajahalme replied:
Added support for local map lookups, but did not factor anything out just yet.

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -182,6 +185,103 @@ type GCFilter struct {
+	}
+	m := bpf.GetMap(mapname)
+	if m == nil {
+		// TODO: Figure out if the global maps ever need to be closed/reopened.
```
I strongly suspect that this is the path that is usually hit, ie `GetMap()` is not providing the ctmap.

AFAIK the ctmap is not always open, it is constantly reopened for garbage collection, and the `mapRegister` will only contain a ctmap while ctmap garbage collection is currently active. In this `m ==nil` case you're explicitly opening the map by path and I don't see how the fd will be released in this path.

In the other path, I'm not so sure about the relationship between the maps cached in `mapRegister` and their corresponding fds. Would need to look into the lifecycle of map registration/unregistration and the corresponding map open/close, and also check whether it's possible for this path to retrieve a reference to a map in memory then attempt to use it while it gets unregistered from another thread.

This is the first usage of `GetMap()` I see in the tree outside metrics so this is worth double-checking. Also, I think that `bpf.OpenMap()` is a lower-level API, usually we would get a `*bpf.Map` object which we invoke `m.Open()` against, which IMO is a bit tidier, in case the `m.Open()` implementation is modified further in future.

Might be worth looking into factoring these changes out into a separate PR so we can closely look at the ctmap open/lookup implementation.

jrajahalme replied:
I'll refactor the ctmap lookup functionality to a separate commit/PR.

jrajahalme replied:
This is still to-be-done.

jrajahalme replied:
Inclined to do that as a follow-on PR.

joestringer commented on pkg/proxy/kafka.go:
```diff
@@ -41,74 +43,166 @@ const (
+	kafkaOnce.Do(func() {
+		kafkaListeners = make(map[uint16]*kafkaListener)
+		kafkaRedirects = make(map[uint64]*kafkaRedirect)
+	})
```
This could probably be simplified and not checked per-redirect creation by just running this in an `func init() { ...}` function right?

jrajahalme replied:
Right, replaced with this:
```
var (
	mutex lock.RWMutex // mutex protects accesses to the configuration resources.
	kafkaListeners = make(map[uint16]*kafkaListener) // key: proxy port
	kafkaRedirects = make(map[uint64]*kafkaRedirect) // key: dst port | dir << 16 | endpoint ID << 32
)
```

joestringer commented on pkg/proxy/kafka.go:
```diff
@@ -41,74 +43,166 @@ const (
+	endpointInfoRegistry logger.EndpointInfoRegistry
+	ingress              bool
+	transparent          bool
+	count                int
```
Might be handy to mention the access patterns here:
* proxySocket() is accessed from multiple goroutines, but that's safe based on `pkg/proxy/socket.go`.
* count is protected by the global mutex below.
* The other parameters are just used by `Listen()` (at least, AFAICT)

jrajahalme replied:
Added this:
```
// The maps holding kafkaListeners (and kafkaRedirects), as well as
// the reference `count` field are protected by `mutex`. `socket` safe
// to be used from multiple goroutines and the other fields below are
// immutable after initialization.
```

joestringer commented on pkg/proxy/kafka_test.go:
```diff
@@ -189,7 +197,24 @@ func (k *proxyTestSuite) TestKafkaRedirect(c *C) {
 
+	// Insert a mock EP to the endpointmanager so that DefaultEndpointInfoRegistry may find
+	// the EP ID by the IP.
+	ep := &endpoint.Endpoint{}
```
FYI there's a new mock endpoint in `pkg/testutils/endpoint.go`, I encourage its use!

jrajahalme replied:
Did not know about it, will have a look!

joestringer commented on pkg/proxy/proxy.go:
```diff
@@ -145,25 +143,115 @@ func (p *Proxy) allocatePort() (uint16, error) {
+	// Sanity-check the proxy port values
+	pmap := make(map[uint16]uint16, 64)
+	for _, v := range ProxyPorts {
+		dscp := v.ProxyPort & 0x3F
```
nit: #define as DSCP_MASK (across everywhere)

jrajahalme replied:
Right, will do!

jrajahalme replied:
This has been since removed as DSCP is not used any more.

joestringer commented on bpf/lib/lxc.h:
```diff
@@ -64,113 +64,55 @@ static inline int is_valid_lxc_src_ipv4(struct iphdr *ip4)
+
+	if (ingress == CT_INGRESS) {
+		// Trace the packet before its forwarded to proxy
+		send_trace_notify(skb, TRACE_TO_PROXY, SECLABEL, 0, 0, HOST_IFINDEX,
```
Why are we only tracing to proxy in the ingress case? Shouldn't we be able to debug knowing that there's forwarding to the proxy on egress case too?

(For both IPv4, IPv6)

jrajahalme replied:
We only "redirect" for ingress, but maybe that should not make a difference.

jrajahalme replied:
I unified these when removing DSCP use.

joestringer commented on bpf/lib/lxc.h:
```diff
@@ -64,113 +64,55 @@ static inline int is_valid_lxc_src_ipv4(struct iphdr *ip4)
+		 * see the modified DSCP. */
+		ipv4_set_dscp(skb, ip4, dscp);
+	} else {
+		skb->mark = MARK_MAGIC_TO_PROXY | dscp;
```
Have you cross-referenced this against the iptables usage in k8s? I'm just a little worried that we're avoiding those lower 8 bits of the mark because kube-proxy uses them for something :worried: 

Either way it might also be useful to add a comment below the `MARK_MAGIC_TO_PROXY` definition stating that the lower bits are often used in this scenario in this way?

jrajahalme replied:
Yes. We use the lowest 8 bits ourselves for a cluster-id, but proxy traffic is always local so it's safe to reuse those bits for this purpose in this case.

joestringer commented on bpf/lib/lxc.h:
```diff
@@ -64,113 +64,55 @@ static inline int is_valid_lxc_src_ipv4(struct iphdr *ip4)
+		 * see the modified DSCP. */
+		ipv4_set_dscp(skb, ip4, dscp);
+	} else {
+		skb->mark = MARK_MAGIC_TO_PROXY | dscp;
```
I would also encourage you to update the https://github.com/cilium/k8s-iptables-diagram , if you haven't done this before then please reach out on slack and I can provide pointers on how to do this.

jrajahalme replied:
Good idea, however will do that once this gets in to avoid back-and-forth with editing the images.

joestringer commented on bpf/lib/ipv4.h:
```diff
@@ -60,4 +60,12 @@ static inline bool ipv4_is_fragment(struct iphdr *ip4)
 
+static inline void ipv4_set_dscp(struct __sk_buff *skb, struct iphdr *ip4, __u8 dscp)
+{
+       	__u8 old_tos = ip4->tos;
```
nit: spacing

jrajahalme replied:
Fixed.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -282,26 +282,12 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
-		cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, HOST_IFINDEX);
-		return redirect(HOST_IFINDEX, 0);
+		// TC_ACT_OK if OK, falling through to the stack.
+		return ipv6_l3(skb, l3_off, (__u8 *) NULL, (__u8 *) &host_mac.addr, METRIC_EGRESS);
```
Looks like we've dropped the `cilium_dbg_capture()` message here, these can be really handy for figuring out what's going on. Can we keep it?

joestringer replied:
In the IPv4 case we keep the DBG_TO_HOST debug trace. I don't know exactly which one we want but let's keep at least one, and keep them consistent between IPv4 and IPv6.

jrajahalme replied:
In this case we do not redirect to `HOST_IFINDEX`, so something has to change?

joestringer replied:
true, maybe `DBG_CAPTURE_DELIVERY` isn't right but the `cilium_dbg(skb, DBG_TO_HOST, skb->cb[CB_POLICY], 0);` is fine. I'd have to double check the monitor side messages for these (probably something like `DbgToHost` and `DbgCaptureDelivery`).

jrajahalme replied:
I have `cilium_dbg_capture(skb, DBG_CAPTURE_PROXY_POST, proxy_port);` in `lib/lxc.c:skb_redirect_to_proxy()`.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -596,30 +582,16 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
-
-		cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, HOST_IFINDEX);
-
-#ifdef HOST_REDIRECT_TO_INGRESS
```
Might need to double-check this against the flannel use case.. /cc @aanm

jrajahalme replied:
Do we have flannel tests in the CI?

joestringer replied:
I think there's a label you can add, then run `test-me-please` to run against flannel. `integration/cni-flannel`

aanm replied:
`test-flannel`

jrajahalme replied:
Flannel tests are currently failing on this PR, apparently the policy routing rule does not match even when the skb->mark is correct (has 0x0200 set), and the packet gets routed back to `cni0` without going to the proxy. Investigating.

tgraf replied:
@jrajahalme Has this been fixed?

jrajahalme replied:
No successful flannel tests on this PR yet :-(

joestringer commented on bpf/lib/common.h:
```diff
@@ -253,7 +253,7 @@ enum {
 #define DROP_CT_MISSING_ACK	-136
 #define DROP_CT_UNKNOWN_PROTO	-137
-#define DROP_CT_CANT_CREATE	-138 /* unused */
+#define DROP_PROXY_INVALID	-138
```
If you're reusing this code for a different error case, please also update the error message in `pkg/monitor/api/drop.go`.

jrajahalme replied:
OK, will do.

jrajahalme replied:
Did not need this after all, so reverted this change.

joestringer commented on bpf/lib/common.h:
```diff
@@ -276,7 +276,7 @@ enum {
 #define DROP_POLICY_L4		-159
 #define DROP_NO_TUNNEL_ENDPOINT -160
-#define DROP_PROXYMAP_CREATE_FAILED	-161
+#define DROP_PROXYMAP_CREATE_FAILED_	-161 /* unused */
```
You might consider adding a comment to `pkg/monitor/api/drop.go` stating that this code is unused.

jrajahalme replied:
Was not aware of the `monitor/api`, will have alook!

jrajahalme replied:
Added a note, also to other unused values. Added a trailing '_' to the unused names to discourage accidental reuse.

joestringer commented on bpf/lib/maps.h:
```diff
@@ -85,28 +85,6 @@ struct bpf_elf_map __section_maps CONFIG_MAP = {
 #endif
 
-#ifdef ENABLE_IPV4
-struct bpf_elf_map __section_maps PROXY4_MAP = {
```
Todo: Search for all remaining references in the codebase. For example, proxy map sizes, map name definitions like this. At least needs updates to `pkg/datapath/linux/config.go`.

jrajahalme replied:
TODO.

jrajahalme replied:
Done (in the next update).

joestringer commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -181,9 +235,15 @@ var ciliumChains = []customChain{
+		removeCiliumRules(t, "iptables")
+	}
+	// Set of tables that has had ip6tables rules in any Cilium version
+	tables6 := []string{"mangle", "raw"}
```
Is this a bug? Why weren't we removing IPv6 rules before?

This commit is huge, _and_ it's iptables :fearful:  Any chance to split it up?

(I'm giving up on the iptables-specific part of the review for now, I can take a fresh look later).

jrajahalme replied:
We weren't adding IPv6 rules before.

Suggestions for how/what to split off are welcome :-)

joestringer replied:
Looking at "bpf: Use iptables TPROXY for proxy redirects ", there's a lot of stuff - total about 2k lines of changes total. I'm guessing that at some point the easiest way to test was squashing everything to this patch, but I see lots of distinct changes:

* IPtables - would be neat if this was a refactor + functional changes in separate patches.
* ctmap changes
* proxymap deletion (We could just unhook the proxymap and leave its code in place, then separately delete the now-unused code)
* The core datapath changes for switching over to tproxy.
* There's maybe one or two other bits but I'm not sure how tightly coupled the changes are across the daemon/datapath

jrajahalme replied:
iptables refactor was split to a separate PR (merged).
Added a new commit to clean up the remaining references to proxymap.
Old proxymap deletion not done yet.
I'm planning for a follow-up PR to refactor the ctmap lookups. Could also do it now if time allows and merge that before this, but that would be dead code, so maybe not.

joestringer commented on pkg/proxy/kafka_test.go:
```diff
@@ -173,6 +179,8 @@ func (k *proxyTestSuite) TestKafkaRedirect(c *C) {
 	defer log.Logger.SetLevel(oldLevel)
 	log.Logger.SetLevel(logrus.DebugLevel)
 
+	flowdebug.Enable()
```
Extraneous flowdebug left in the commit?

jrajahalme replied:
Right, removed.

joestringer commented on test/provision/k8s_install.sh:
```diff
@@ -275,4 +275,7 @@ docker network create --subnet=192.168.9.0/24 outside
 docker run --net outside --ip 192.168.9.11 --restart=always -d docker.io/cilium/demo-httpd:latest
 
+#Preload the ip6tables table we need
+sudo modprobe ip6table_mangle
```
Do we need this elsewhere (rather than just ginkgo CI provision)? Why isn't it autoloaded? Should we document it? Should we put it in the VM images in github.com/cilium/packer-ci-build? Container images?

jrajahalme replied:
This will definitely need more discussion and a solution that works everywhere. It is not autoloaded because the cilium container has no privileges to do so (CAP_NET_ADMIN is not enough for kernel module autoloading).

ianvernon replied:
@jrajahalme does what you say mean that this may not work in existing Cilium setups upon upgrade?

jrajahalme replied:
This issue needs still to be addressed. My preferred solution would be to give cilium container the prerequisite capability to (auto-) load kernel modules. @tgraf ??

tgraf replied:
I'm not opposed to adding `CAP_SYS_MODULE` but we need to document why we need it and when it can be removed.

joestringer commented on test/helpers/cons.go:
```diff
@@ -28,7 +28,7 @@ import (
 var (
 	// HelperTimeout is a predefined timeout value for commands.
-	HelperTimeout = 5 * time.Minute
+	HelperTimeout = 8 * time.Minute
```
:(

jrajahalme replied:
Hopefully the CI boxes don't need to rely in this, this was needed for my local ginko runs.

ianvernon replied:
This PR should not be merged with this change. If it takes longer than 5 minutes for pods to initialize, then the cause should be determined.

jrajahalme replied:
I'll try leaving it out. As said it was needed for local ginkgo run, and my internet connection may have been slower than normal at the time, etc.

joestringer reviewed (changes requested):
Wow, this is an epic overhaul.

I've tried my best to do a first-run through. I haven't looked closely at the iptables section of this, partly because that commit was huge and there's both functional and refactoring changes in one giant patch with changes across many packages.

I noticed multiple places where there's code that looks "a bit weird", where I ask "Is this a bug?". In those cases it'd be good to get to the bottom of it and determine whether there is a necessity for smaller bugfixes that can be applied first to the tree and backported.

One concern I have is in the new usage of the mark, maybe you already cross-referenced against k8s usage of mark values in iptables and it's OK, but I wanted to ask about it anyway.

There's a few remaining proxymap references in the codebase, and there's some monitor messages that could do with some updates.

It would be nice to factor out the ctmap changes into a separate PR that can be looked at separately with closer eyes.

I haven't looked at the kafka code before, it looks pretty straightforward to me so hopefully I'm not missing anything subtle there.

There's lots of tasty removal bits here of ugly proxymap code, both on the kernel side and user side so I'm pretty excited about this. It's pretty awesome that you're close to a net even LoC change, suggests that the functionality is being streamlined quite nicely. 

I recognise that this corresponds with https://github.com/cilium/proxy/pull/11, but I haven't attempted review of that as I'm entirely unfamiliar with that area. I'm willing to take a look if there isn't someone more suited to review that but I won't guarantee how much I'll notice :-)

As always, please filter through my comments and address the ones you think are important for the base merge and consider queueing up less significant feedback into an issue which can be handled by a follow-up PR. This is a lot of work and I can appreciate that attempting to rework nits in older patches is a pain.

aanm commented on Dockerfile:
```diff
@@ -1,7 +1,7 @@
 # cilium-envoy from github.com/cilium/proxy
 #
-FROM quay.io/cilium/cilium-envoy:84ee839e1d78ef858a39e390288ad417d35bb1d4 as cilium-envoy
+FROM quay.io/cilium/cilium-envoy:9e1a3317863a5dd0e393fd168ce85cc2a3c709e4 as cilium-envoy
```
Please also update https://github.com/cilium/cilium/blob/95fdd49282abda22991f224d9e9e249e882d7650/Gopkg.toml#L54-L56

jrajahalme replied:
These two SHAs may sometimes be out of sync. The Godep dependency is for the API (protobufs), and it may not need to be updated everytime something is fixed in the proxy. Dockerfile refers to the cilium-envoy build and needs to change when bugs are fixed or new features are added to the proxy.

aanm commented on pkg/proxy/proxy.go:
```diff
@@ -212,19 +212,30 @@ func (p *Proxy) CreateOrUpdateRedirect(l4 *policy.L4Filter, id string, localEndp
+	var to uint16
+	fixedProxyPort := false
+	listenerName := id
+	// Detect the need for a reusable listener here and override the defaults, e.g.:
```
Is this code suppose to be commented?

jrajahalme replied:
The commented code was removed in a later commit. I have cleaned up the commits so that an issue like this does not exist any more.

aanm commented on pkg/proxy/proxy.go:
```diff
@@ -222,6 +222,17 @@ func (p *Proxy) CreateOrUpdateRedirect(l4 *policy.L4Filter, id string, localEndp
+	if l4.L7Parser == policy.ParserTypeHTTP {
+		if l4.Ingress {
+			listenerName = "cilium-http-ingress"
+			to = 6773
```
I'm a little lost, where did this 6773 and 6769 came from?

jrajahalme replied:
They were fixed port numbers I just came up with for testing. The PR has since been moved to dynamically allocate the proxy ports.

aanm commented on pkg/proxy/kafka.go:
```diff
@@ -41,12 +44,38 @@ const (
+var kafkaListeners map[uint16]*kafkaListener // key: proxy port
+var kafkaRedirects map[uint64]*kafkaRedirect // key: dst port | ingress << 16 | endpoint ID << 32
+
+func mapKey(dstPort uint16, ingress bool, eID uint64) uint64 {
```
to avoid making the wrong assumptions it would be nice to make `eID` a `uint32`, otherwise ` mapKey(1, true, 0x900000001)` and `mapKey(1, true, 0x1)` will return the same key.

jrajahalme replied:
I did this change in a local commit not pushed to this PR yet, will push it with the next update.

aanm commented on pkg/proxy/kafka.go:
```diff
@@ -41,12 +44,38 @@ const (
+	return uint64(dstPort) | dir<<16 | eID<<32
+}
+
+var kafkaOnce sync.Once
```
Please add documentation of what this `kafkaOnce` is for or I'm forced to make a joke... kafkaOnce, kakfaTwice, sold!

jrajahalme replied:
You are having too much of a good time reviewing this PR, so I removed `kafkaOnce` :-)

aanm commented on pkg/proxy/kafka.go:
```diff
@@ -69,45 +161,54 @@ func createKafkaRedirect(r *Redirect, conf kafkaConfiguration, endpointInfoRegis
+	mutex.Lock()
+	if _, ok := kafkaRedirects[key]; ok {
+		mutex.Unlock()
+		panic("Kafka redirect already exists for the given dst port and endpoint ID")
```
let's not panic here since, as far as I understand, this `createKafkaRedirect` can be called by the same `r.dstPort, r.ingress, r.endpointID` tuple at the "same time". For example 2 redirects happening at the same time, the first redirect calls this function which sets this up, the 2nd redirect also calls this function and will wait on line 166 until the mutex is released by the 1rst redirect setup. Once the mutex is released the 2nd redirect will panic because the 1st redirect had already configure this

jrajahalme replied:
Higher level mutex in `CreateOrUpdateRedirect()` protects against this from happening.

aanm commented on pkg/proxy/redirect.go:
```diff
@@ -52,6 +52,7 @@ type Redirect struct {
 	// ProxyPort is the port the redirects redirects to where the proxy is
 	// listening on
 	ProxyPort      uint16
+	dstPort        uint16
```
can you add documentation WRT `dstPort`?

jrajahalme replied:
Done for the next update (its the L4 port the redirect is for, e.g., 80).

aanm commented on pkg/proxy/proxy.go:
```diff
@@ -213,24 +213,42 @@ func (p *Proxy) CreateOrUpdateRedirect(l4 *policy.L4Filter, id string, localEndp
+	case policy.ParserTypeKafka:
+		if l4.Ingress {
+			listenerName = "cilium-kafka-ingress"
+			to = 6774
```
it will be useful if all of these ports were a constants on top of this file with documentation. Will we fail hard if we can't allocate all of those continuous ports?

jrajahalme replied:
Proxy ports are now dynamically allocated.

aanm commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -466,39 +470,32 @@ func ExtractMsgDetails(msg *dns.Msg) (qname string, responseIPs []net.IP, TTL ui
 	bindAddr := net.JoinHostPort(address, strconv.Itoa(int(port)))
 
-	TCPAddr, err := net.ResolveTCPAddr("tcp", bindAddr)
+	listener, err = listenConfig(0xb00).Listen(context.Background(), "tcp", bindAddr)
```
please add documentation for the meaning of `0xb00`

jrajahalme replied:
Did not mean to spook you :-)

Replaced with `linux_defaults.MagicMarkEgress`

aanm commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -466,39 +470,32 @@ func ExtractMsgDetails(msg *dns.Msg) (qname string, responseIPs []net.IP, TTL ui
-	}
-	UDPConn, err = net.ListenUDP("udp", UDPAddr)
+
+	conn, err = listenConfig(0xb00).ListenPacket(context.Background(), "udp", listener.Addr().String())
```
please add documentation for the meaning of `0xb00`

jrajahalme replied:
ditto

aanm commented on pkg/fqdn/dnsproxy/udp.go:
```diff
@@ -0,0 +1,250 @@
+package dnsproxy
```
missing copyright header

jrajahalme replied:
Added.

aanm commented on pkg/fqdn/dnsproxy/udp.go:
```diff
@@ -0,0 +1,250 @@
+	//   checking that a route exists from the source address before
+	//   the source address is replaced with the (transparently) changed one
+	udpOnce.Do(func() {
+		rawconn4 = bindUDP("127.0.0.1") // raw socket for sending IPv4
```
same thing that joe said, you can initialize this directly in this file

jrajahalme replied:
Done.

aanm commented on pkg/fqdn/dnsproxy/udp.go:
```diff
@@ -0,0 +1,250 @@
+	s.raddr = nil
+	s.laddr = nil
+	s.m = s.m[:cap(s.m)]
+	s.oob = s.oob[:cap(s.oob)]
```
I don't understand this.

1) how can you guarantee the slice hasn't increased its capacity?
2) I think this needs to be at least `make([]byte, udpOOBSize, cap(s.oob))`. Similar above `make([]byte, len(s.m) ,msgSize)` so we are sure the slice is really "clean" up

jrajahalme replied:
The above changes the length of the slice to be the same as it's capacity. syscall.Recvmsg() gets the oob slice and only uses it upto the length of it.

Initialization `make([]byte, udpOOBSize)` creates a new slice with the given capacity, initializes the elements to zero values and returns with `len() == udpOOBSize`. As such the slice is ready to be used with `syscall.Recvmsg()`. After the UDP IO we re-slice the data to the actual number of bytes returned by the syscall, so then `len() <= cap()`. Syscall never increases the capacity.

In `Clear()` (quoted above), we re-slice `oob` (respecitively, `m`) so that `len()` is again the same as the full capacity (`cap()`). The contents need not be re-initialized, these are read buffers.

So:
1. Yes, and even if it did, it should not matter. 
2. It's better to avoid reallocating the memory.

aanm commented on pkg/fqdn/dnsproxy/udp.go:
```diff
@@ -0,0 +1,250 @@
+	binary.Write(bb, binary.BigEndian, uint16(s.laddr.Port))
+	binary.Write(bb, binary.BigEndian, uint16(s.raddr.Port))
+	binary.Write(bb, binary.BigEndian, uint16(8+len(b)))
+	binary.Write(bb, binary.BigEndian, uint16(0)) // checksum
```
I like this checksum :-)

jrajahalme replied:
AFAIK kernel will fill it in.

aanm commented on pkg/fqdn/dnsproxy/udp.go:
```diff
@@ -0,0 +1,250 @@
+		n, _, err = rawconn4.WriteMsgIP(buf, s.controlMessage(s.laddr), &dst)
+	}
+	if err != nil {
+		log.Fatalf("WriteMsgIP: %s", err)
```
Why the fatal?

jrajahalme replied:
Remnant from debugging, thanks for pointing this out :-)

aanm commented on pkg/fqdn/dnsproxy/udp.go:
```diff
@@ -0,0 +1,250 @@
+			// Address family is in native byte order
+			family := *(*uint16)(unsafe.Pointer(&msg.Data[unsafe.Offsetof(pp.Family)]))
+			if family != unix.AF_INET {
+				return nil, fmt.Errorf("original destination is not IP.")
```
you mean `original destination is not IPv4.`?

jrajahalme replied:
Yes, thanks!

aanm commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -182,6 +185,103 @@ type GCFilter struct {
+	if ingress {
+		key.Flags = TUPLE_F_IN
+	}
+	return key, nil
```
Instead of setting the sport and dport has network byte order, you could set it as host network byte order and then here you could call `CtKey.ToNetwork()`

jrajahalme replied:
Right, changed!

aanm reviewed (dismissed):
I've reviewed per commit so some of my comments might be outdated

ianvernon commented on pkg/endpoint/endpoint.go:
```diff
@@ -885,6 +885,13 @@ func (e *Endpoint) HasSidecarProxy() bool {
 	return e.hasSidecarProxy
 }
 
+func (e *Endpoint) ConntrackName() string {
```
Any need to hold the endpoint lock while this is being performed? I doubt it, but I see that this uses `ConntrackLocalLocked` which assumes the endpoint is locked for reading.

joestringer replied:
it's using the options of the endpoint to determine whether conntrack is local for this endpoint, which can be changed by the API asynchronously right?

jrajahalme replied:
In practice most call sites already have the endpoint (read) locked where this is used.

tgraf replied:
@jrajahalme Please clarify the locking requirement in the comment of the function.

jrajahalme replied:
Done!

ianvernon commented on Dockerfile:
```diff
@@ -1,7 +1,7 @@
 #
```
Add some context to this commit as to why this is necessary - it's helpful to get an overall summary when looking through the changelog / reviewing.

jrajahalme replied:
Added this:
"Use newer proxy image that has support for deriving policy name from
the endpoint IP, and for deriving the proxylib parser name from the
network policy.

These changes allow listener configuration to be independent of an
endpoint or an endpoint's network policy."

ianvernon commented on bpf/lib/maps.h:
```diff
@@ -85,28 +85,6 @@ struct bpf_elf_map __section_maps CONFIG_MAP = {
 #endif
 
-#ifdef ENABLE_IPV4
-struct bpf_elf_map __section_maps PROXY4_MAP = {
```
(not sure on this, which is why this is a question - maybe @joestringer can answer as well) - will these maps get cleaned up automatically now that they are removed from the Cilium codebase, or will they continue to reside on the BPF filesystem?

joestringer replied:
Good point, we should add a unit test to `pkg/datapath/maps` to ensure it's cleaned up.

jrajahalme replied:
I have not addressed this in this PR.

ianvernon reviewed (commented):
A few misc. comments. I don't feel particularly qualified to review most of this this because I'm not sure of the implication some of the changes for which I am a codeowner (i.e., to endpoint) have on the datapath / vice versa :/ .

tgraf commented on Gopkg.toml:
```diff
@@ -295,7 +295,8 @@ required = [
   name = "github.com/miekg/dns"
-  revision = "v1.1.4"
+  revision = "164b22ef9acc6ebfaef7169ab51caaef67390823"
+  source = "github.com/cilium/dns"
```
Did we upstream this?

jrajahalme replied:
I proposed this in [an upstream PR](https://github.com/miekg/dns/pull/918,) but they countered with a less invasive approach. The good thing is that the need for upstream changes was acknowledged. It may take some more time to finalize, but IMO it is safe to merge with what we have now, and then update back to the upstream branch when that is ready.

tgraf commented on pkg/datapath/fake/datapath.go:
```diff
@@ -58,3 +58,11 @@ func (f *fakeDatapath) WriteNetdevConfig(io.Writer, datapath.DeviceConfiguration
 	return nil
 }
+
+func (l *fakeDatapath) InstallProxyRules(uint16, bool, string) error {
```
Please fix these

jrajahalme replied:
Fixed!

tgraf commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -233,6 +250,179 @@ func RemoveRules() {
+		"-p", l4Match,
+		"-m", "mark", "--mark", markMatch,
+		"-m", "comment", "--comment", "cilium: TPROXY to host " + name + " proxy",
+		"-j", "TPROXY",
```
This is a new requirement that we need to document in the system requirements. How common is TPROXY support in distribution configurations?

jrajahalme replied:
Both TPROXY and NOTRACK are widely supported (#7435)

tgraf commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -233,6 +250,179 @@ func RemoveRules() {
+			"-m", "iprange", "--dst-range", iptRange(node.GetIPv4AllocRange()),
+			"-m", "mark", "--mark", matchToProxy,
+			"-m", "comment", "--comment", "cilium: NOTRACK for proxy traffic",
+			"-j", "NOTRACK"}, false)
```
NOTRACK is definitely not widely available. What will break if this is not available?

tgraf commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -182,6 +185,102 @@ type GCFilter struct {
+		return nil, fmt.Errorf("unable to parse port string: %s", err)
+	}
+
+	local_ip, local_port, err := net.SplitHostPort(localAddr)
```
Please fix these

jrajahalme replied:
Done!

tgraf reviewed (changes requested):
This is super exciting at a high level. The main question is the availability of TPROXY and NOTRACK. The kernel module autoloading must be fixed before we can merge this.

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -181,6 +183,106 @@ type GCFilter struct {
+
+	m := bpf.GetMap(mapname)
+	if m == nil {
+		// TODO: Figure out if the global maps ever need to be closed/reopened.
```
This TODO is still here, I'm not sure what the status is for the thread that I previously opened... will repeat here because finding the original comment is almost impossible on github:

I strongly suspect that this is the path that is usually hit, ie `GetMap()` is not providing the ctmap.

AFAIK the ctmap is not always open, it is constantly reopened for garbage collection, and the `mapRegister` will only contain a ctmap while ctmap garbage collection is currently active. In this `m ==nil` case you're explicitly opening the map by path and I don't see how the fd will be released in this path.

In the other path, I'm not so sure about the relationship between the maps cached in `mapRegister` and their corresponding fds. Would need to look into the lifecycle of map registration/unregistration and the corresponding map open/close, and also check whether it's possible for this path to retrieve a reference to a map in memory then attempt to use it while it gets unregistered from another thread.

This is the first usage of `GetMap()` I see in the tree outside metrics so this is worth double-checking. Also, I think that `bpf.OpenMap()` is a lower-level API, usually we would get a `*bpf.Map` object which we invoke `m.Open()` against, which IMO is a bit tidier, in case the `m.Open()` implementation is modified further in future.

Might be worth looking into factoring these changes out into a separate PR so we can closely look at the ctmap open/lookup implementation.

jrajahalme replied:
My intent was to use the existing map registry to keep the ctmap open, so that each lookup would not need to bear the cost of opening and closing it, but this really works for the global map only, as this leaks fds for removed endpoints with local ct maps.

I've now changed this to:
- Add this new lookup functionality to `pkg/maps/ctmap/lookup.go`
- Add new `CloseLocalMaps()` that is called from `endpoint.LeaveLocked()` to close any opened local conntrack maps.

I don't readily see how to use the existing `maps()` function et al to get the protocol specific map. I'd like to leave tidying this up further for a follow-on PR.

joestringer replied:
Ah, so:
* At least one lookup is guaranteed to happen,
* Furthermore the first lookup is guaranteed to open the map and leave it open,
* Then Close is guaranteed to remove a reference to the map.

And for global maps this will hopefully just open exactly one reference and for the local maps they won't be shared so each endpoint manages the reference to the fd in that case.

Seems fine at a glance, but perhaps a touch brittle. I'd welcome a follow-on PR to tidy this up a bit and make stronger guarantees here. Either way, I'm sure we'll find out sooner or later if the fd logic is wrong through fd open stats in testing clusters. SGTM for now. (Would also welcome filing an issue to highlight the potential tech debt here and manage, if you aren't likely to work on it soon)

joestringer commented on bpf/lib/ipv4.h:
```diff
@@ -60,4 +60,12 @@ static inline bool ipv4_is_fragment(struct iphdr *ip4)
 	return ip4->frag_off & bpf_htons(0xBFFF);
 }
 
+static inline void ipv4_set_dscp(struct __sk_buff *skb, struct iphdr *ip4, __u8 dscp)
```
Now-unused function? (Similarly for ipv6)

jrajahalme replied:
Removed, thanks!

joestringer commented on bpf/sockops/bpf_redir.c:
```diff
@@ -61,39 +61,10 @@ int bpf_redir_proxy(struct sk_msg_md *msg)
-			return SK_PASS;
-		err = msg_redirect_hash(msg, &SOCK_OPS_MAP, &key, flags);
-	} else if (!verdict) {
+	if (verdict != 0) {
```
Comparing the previous version of this vs the current version of this, I'm confused.

Previously:

```
if (redirect_to_proxy(verdict)) {  // ie, if verdict is > 0
   ... update proxy4map
   ... msg_redirect_hash(...)
} else if (!verdict) { // ie, verdict is 0
   ... msg_redirect_hash(...)
}
```

Now:

```
if (verdict != 0) { // ie, < 0 or > 0
   ...
}
```

jrajahalme replied:
Thanks for spotting this, changed this to:
```
	if (verdict >= 0) {
		msg_redirect_hash(msg, &SOCK_OPS_MAP, &key, flags);
	}
```

This should retain previous functionality.

joestringer replied:
Is there some tests we can run against this? Such a test should have shown the failure previously and show that it now works (unless there's something else hiding in the wings)

joestringer commented on daemon/daemon.go:
```diff
@@ -550,15 +550,17 @@ func (d *Daemon) compileBase() error {
-				return err
-			}
+	// Always remove masquerade rule and then re-add it if required
+	iptables.RemoveRules()
```
Apologies, I haven't filtered through the commit messages, but this was previously only done if IPv4 is enabled. Is this deliberate?

jrajahalme replied:
Yes. Previously all rules were IPv4 rules. Now some are IPv6 rules. Also, RemoveRules() can't fail, so no new error conditions are added. Besides, this is a little bit more correct now, as it is possible that node configuration has changed (the thing being processed here) so that IPv4 is changed from enabled to disabled. The old IPv4 rules should be removed also in that case.

`iptables.InstallRules()` now installs IPv4 rules only if IPv4 is enabled, and ditto for IPv6, so call to it should not be gated by IPv4 being enabled any more.

joestringer reviewed (changes requested):
I see a few remaining key points:
* Does https://github.com/cilium/proxy/pull/11 need to be merged first?
* Question about fd leak during ctmap lookup
* Possible flannel breakage
* Is sockmap broken by this? (New comment below)
* Removal of old `proxymap` maps from filesystem (after all endpoint datapaths are migrated over? I see this is deferred for now and maybe for upgrade/downgrade purposes it would be simpler to leave in place but maybe we should file an issue to ensure these are removed in release N+1)

I didn't do a thorough re-review, mostly went back over previous feedback and took a quick glance through BPF side again. The datapath changes look a bit cleaner this time, particularly wrt monitor messages.

ianvernon commented on pkg/proxy/kafka.go:
```diff
@@ -46,7 +46,6 @@ type kafkaRedirect struct {
 	redirect             *Redirect
 	endpointInfoRegistry logger.EndpointInfoRegistry
 	conf                 kafkaConfiguration
-	rules                policy.L7DataMap
```
Glad to see policy package implementation no longer being here!

jrajahalme committed:
Dockerfile: Update proxy dependency

Use newer proxy image that has support for deriving policy name from
the endpoint IP, and for deriving the proxylib parser name from the
network policy.

These changes allow listener configuration to be independent of an
endpoint or an endpoint's network policy.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
CI: Change Kafka runtime tests to use local conntrack maps.

It's good to have test coverage with the local conntrack maps.
Optimally would keep also the global map test cases.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
redirect: rename 'id' as 'listenerName'

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
Envoy: Do not configure policy name

When left empty, the POD IP is used as the policy name. We already
used the IPv4 address of the POD as the policy name. Now it can also
be the IPv6 address, but we already send the policy down with both
IPv4 and IPv6 as names.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
kafka: Remove unused field.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
endpointmanager: Add LookupIP()

LookupIP() finds the local endpoint given the IP address of the
endpoint in the string form.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxy: Add CT map name to the network policy to support local CT maps.

Add conntrack map name to the network policy passed to
cilium-envoy. When specified, conntrack maps are used for source
security identity lookup instead of the proxymaps. "global" should be
specified for the global CT maps. The full bpf path is constructed
based on this map name using the bpf path configuration option of the
cilium bpf_metedata filter and the CT map naming conventions.

This options allows the network policy of a specific endpoint to
specify which conntrack maps must be used to retrieve the metadata for
the incoming connections. If the endpoint is configured to use local
conntrack maps, cilium-agent will use this new 'conntrack_map_name' to
pass the endpoint ID as a string allowing the proxy to form the full
paths to the endpoint's local conntrack maps.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
fqdn: Adapt to TPROXY

Make dnsproxy to work with TPROXY redirection, where the destination
address and port are not NATted.

Make DNS proxy unit tests privileged (needed for the socket options &
mark).

Do not start the DNS proxy when running in the (unprivileged) DryMode.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
vendor: Use cilium/dns for miekg/dns, Use extended SessionUDP

Use a SessionUDP with TPROXY support:
- Set IP_TRANSPARENT and IP_RECVORIGDSTADDR options
- Use raw UDP sockets (IPv4 & IPv6) for sending responses, so that
  they can be sent from the same source port as the corresponding
  requests were sent to without requiring to create and bind a new
  socket for each possible source port.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
bpf: Use iptables TPROXY and shared proxy listeners

Change all Cilium proxies to use transparent redirect (TPROXY) to
shared proxy ports. This allows IP addresses and ports to remain
unchanged on redirected packets and removes the need to use a bpf map
to get the original destination address and port. Thus the proxymap is
completely removed, which helps reduce complexity of the bpf datapath.

Since the bpf socket helper is not yet available, we insert a set of
iptables rules for proxy redirects after a proxy has been configured.

Proxies still need to get the source security ID. This is done with a
lookup in the Cilium conntrack bpf maps. Lookup support is added for
both local and global CT maps.

Linux conntrack interferes with TPROXY as conntrack can NAT the traffic
due to 5-tuple collisions while TPROXY may use the non-NATted 5-tuple
for socket matching. This can lead to TCP 3-way handshake never finishing
due to TPROXY directing all packets to the listening socket, instead
of associating non-SYN packets with the new socket. This is resolved
by adding NOTRACK rules for proxy traffic.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
